### PR TITLE
Adding semantic versioning and caching of binary

### PR DESCRIPTION
**Pull request by dkcumming**
Adding [semantic versioning](https://semver.org/), and caching the nix built binary on master push.

### CODE REVIEW

**Comment by dkcumming** [comment]
@F-WRunTime I am not the best at dev ops. Does [this mean](https://github.com/runtimeverification/mir-semantics/actions/runs/6965359332/job/18953918686#step:4:10) I need to split this into two PR's, one to add the `package/version`, and then the next with the job?

File: package/version.sh [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    current_version_minor="$(echo ${current_version} | cut --delimiter '.' --field 2)"
+    current_version_patch="$(echo ${current_version} | cut --delimiter '.' --field 3)"
+    new_version="${version}"
+    if [[ "${version_major}" == "${current_version_major}" ]] && [[ "${version_minor}" == "${current_version_minor}" ]]; then
+        new_version="${version_major}.${version_minor}.$((version_patch + 1))"
+    fi
+    # If the file being commited increases the major or minor, then take that version as it is
+    if [[ "${version_major}" < "${current_version_major}" ]] || [[ "${version_minor}" < "${current_version_minor}" ]]; then
+        new_version="${current_version}"
+    fi
```
Comment (dkcumming) [inline package/version.sh:27]: @F-WRunTime I want us to have the ability to increase the major and minor by updating `package/version`, does this seem reasonable?

File: kmir/src/kmir/__init__.py [hunk 2]
```diff
@@ -1 +1,5 @@
+from typing import Final
+
 from .kmir import KMIR
+
+VERSION: Final = '0.2.0'
```
[the remark this answers is not in the document]
Comment (tothtamas28) [reply kmir/src/kmir/__init__.py:5]: In `evm-semantics`: command `kevm version`
https://github.com/runtimeverification/evm-semantics/blob/8c169c89ebf4e3541d93bad16e835aed85cc0e79/kevm-pyk/src/kevm_pyk/__main__.py#L87

In `kontrol`: command `kontrol version`
https://github.com/runtimeverification/kontrol/blob/91ebc7175eb19228a35e120d54e4fb56a6f55e63/src/kontrol/__main__.py#L104

File: package/version.sh (hunk 1 above)
Comment (F-WRunTime) [reply package/version.sh:27]: Yes this looks good. If you want to rev the major version this is a manual step on your part, the version.sh script will ignore rev'ing the minor version in this scenario.

**Review by F-WRunTime — APPROVED** [review]

**Comment by F-WRunTime** [comment]
> @F-WRunTime I am not the best at dev ops. Does [this mean](https://github.com/runtimeverification/mir-semantics/actions/runs/6965359332/job/18953918686#step:4:10) I need to split this into two PR's, one to add the `package/version`, and then the next with the job?

This is failing because the reference branch master does not have the file `package/version` this is fine. Yeah it would be cleaner to see everything passing and we can adjust the version bump in a second PR.

File: kmir/src/kmir/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dkcumming) [reply kmir/src/kmir/__init__.py:5]: Ah actually that is something we want. Thank you.

File: .github/workflows/master-push.yml [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        run: |
+          set -x
+          VERSION=$(cat package/version)
+          curl --fail                                                          \
+            -X POST                                                            \
+            -H "Accept: application/vnd.github+json"                           \
+            -H "Authorization: Bearer ${{ secrets.JENKINS_GITHUB_PAT }}"       \
+            -H "X-GitHub-Api-Version: 2022-11-28"                              \
+            https://api.github.com/repos/runtimeverification/devops/dispatches \
+            -d '{"event_type":"on-demand-test","client_payload":{"repo":"runtimeverification/kmir","version":"'${VERSION}'"}}'
```
[the remark this answers is not in the document]
Comment (dkcumming) [reply .github/workflows/master-push.yml:110]: I don't think there are any dependants that it is necessary to keep this around.

It is very unlikely that we will be updating `mir-semantics-compiletest` anytime soon. The tests we have already form reasonable coverage for 2021 edition of Rust.

**Pull request merged**